### PR TITLE
Implement IO port API

### PR DESCRIPTION
**Pull request by LaszloLango**
Related issue: #964

JerryScript-DCO-1.0-Signed-off-by: László Langó llango.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by LaszloLango** [comment]
@pfalcon, @sergioamr  please check the Zephyr port. I hope my speculative update work as it should.

**Comment by LaszloLango** [comment]
Note: after this PR, we will change all the printf calls inside the engine to the related port API function (`jerry_port_log` and `jerry_port_console`)

**Comment by LaszloLango** [comment]
cc: @jiangzidong

File: jerry-core/jrt/jrt.h [hunk 1]
```diff
@@ -109,7 +109,7 @@ extern void __noreturn jerry_unimplemented (const char *, const char *, const ch
 #define JERRY_DDDLOG(...) JERRY_DLOG (__VA_ARGS__)
 #endif /* JERRY_ENABLE_LOG */
 
-#define JERRY_ERROR_MSG(...) jerry_port_errormsg (__VA_ARGS__)
+#define JERRY_ERROR_MSG(...) jerry_port_log (JERRY_LOG_LEVEL_ERROR, __VA_ARGS__)
 #define JERRY_WARNING_MSG(...) JERRY_ERROR_MSG (__VA_ARGS__)
```
Comment (galpeter) [inline jerry-core/jrt/jrt.h]: Shouldn't the warning use warning level?

**Comment by sergioamr** [comment]
Will check it in my morning tomorrow :)

**Comment by jiangzidong** [comment]
It will bring convenience to porting work. Thanks

File: targets/esp8266/user/jerry_port.c [hunk 2]
```diff
@@ -17,34 +17,36 @@
...
+void
+jerry_port_console (const char *format, /**< format string */
+                    ...) /**< parameters */
 {
   va_list args;
-  int count = 0;
   va_start (args, format);
   // TODO, uncomment when vfprint link is ok
-  //count = vfprintf (stream, format, args);
+  // vfprintf (stdout, format, args);
```
Comment (zherczeg) [inline targets/esp8266/user/jerry_port.c]: please use /* */ comment style

File: targets/esp8266/user/jerry_port.c [hunk 3]
```diff
@@ -17,34 +17,36 @@
...
+                ...)  /**< parameters */
 {
+  (void) level; /* ignore log level */
+
   va_list args;
-  int count = 0;
   va_start (args, format);
   // TODO, uncomment when vprint link is ok
-  //count = vprintf (format, args);
+  // vprintf (stderr, format, args);
```
Comment (zherczeg) [inline targets/esp8266/user/jerry_port.c]: Ditto

**Comment by zherczeg** [comment]
LGTM after the comment fixes.

**Comment by dbatyai** [comment]
LGTM

File: jerry-core/jrt/jrt.h [hunk 4]
```diff
@@ -86,9 +86,9 @@ extern void __noreturn jerry_unimplemented (const char *, const char *, const ch
 #define JERRY_LOG(lvl, ...) \
   do \
   { \
-    if (lvl <= jerry_debug_level && jerry_log_file) \
+    if (lvl <= jerry_debug_level) \
```
Comment (akosthekiss) [inline jerry-core/jrt/jrt.h]: IMHO, this check should not happen here but in the port implementation (if that impl cares about it). Likewise, `jerry_debug_level` shouldn't even be defined in jerry.c. Everything `JERRY_LOG` should do is just pass things on to `jerry_port_log` implemented in the port for it to decide what to do with it (as `lvl` is passed on as well).

In some sense, it should work similarly to the "fatal" logic. As far ar jerry-core is concerned, there is only a single `jerry_port_fatal` function to be called. However, the default port implementation adds some extras to it, namely the "abort_on_fail" functionality. However, that's a port-specific thing. Therefore, the `static bool abort_on_fail` variable is defined in jerry-port-default-fatal.c, and accessor methods `jerry_port_default_set_abort_on_fail` and `jerry_port_default_is_abort_on_fail` in jerry-port-default.h.

I can imagine a similar scenario with the logs. There could be default port-specific accessors, e.g., `jerry_port_default_set_log_level` and `jerry_port_default_get_log_level` in jerry-port-default.h and their implementation (along with any variables needed for that logic, say `debug_level` or `log_level`) in jerry-port-default-io.c. And then the line `(void) level; /* default port implementation ignores the log level */` will go away (actually, the presence of this line already signals the problem that log levels are not checked at the right place, IMHO).

Additonal notes:
* `JERRY_D[D[D]]LOG` macros call `JERRY_LOG` with 1, 2, and 3 as level parameters. This does not seem right in the light of the new log level enum.
* There are still some `printf`s in the code (e.g., jrt-fatals.c, js-parser.c, js-parser-util.c, ecma-builtin-global.c, jmem-poolman.c, jmem-heap.c, js-parser-statm.c), even if mostly macro-guarded. When will these be rewritten to call the `_log` and `_console` functions?

Comment (akosthekiss) [inline jerry-core/jrt/jrt.h]: (Sorry, missed the "Note: after this PR," part. Cancel my last Q.)

**Comment by sergioamr** [comment]
tested branch on my arduino101. LGTM

File: jerry-core/jrt/jrt.h (hunk 4 above)
Comment (LaszloLango) [inline jerry-core/jrt/jrt.h]: I disagree with the first part. `JERRY_D[D[D]]LOG` macros are all on debug level. There where no multiple debug levels in the proposed (and accepted) IO port API proposal (see: #964).

I think moving the debug level to the port is might be a good idea. I'll try to do it in a followup work.

**Comment by LaszloLango** [comment]
@sergioamr, thanks

File: jerry-core/jrt/jrt.h (hunk 4 above)
Comment (akosthekiss) [inline jerry-core/jrt/jrt.h]:
```c
#ifdef JERRY_ENABLE_LOG
#define JERRY_LOG(lvl, ...) \
  do \
  { \
    if (lvl <= jerry_debug_level) \
    { \
      jerry_port_log (JERRY_LOG_LEVEL_DEBUG, __VA_ARGS__); \
    } \
  } \
  while (0)

#define JERRY_DLOG(...) JERRY_LOG (1, __VA_ARGS__)
#define JERRY_DDLOG(...) JERRY_LOG (2, __VA_ARGS__)
#define JERRY_DDDLOG(...) JERRY_LOG (3, __VA_ARGS__)
#else /* !JERRY_ENABLE_LOG */
#define JERRY_DLOG(...) \
  do \
  { \
    if (false) \
    { \
      jerry_ref_unused_variables (0, __VA_ARGS__); \
    } \
  } while (0)
#define JERRY_DDLOG(...) JERRY_DLOG (__VA_ARGS__)
#define JERRY_DDDLOG(...) JERRY_DLOG (__VA_ARGS__)
#endif /* JERRY_ENABLE_LOG */
```
This is current status. `JERRY_D[D[D]]LOG` macros expand to `JERRY_LOG (]0-3], ...)` depending on `JERRY_ENABLE_LOG` settings. The first parameter should be the debug level you just mentioned, not integers.

Comment (LaszloLango) [inline jerry-core/jrt/jrt.h]: @akosthekiss, as I said. I'll do it in a follow up work. Don't worry. It's on my list to clean this up.

**Pull request merged**